gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientLibraryWriteChannel.java

```diff
@@ -0,0 +1,209 @@
+            .setContentEncoding(createOptions.getContentEncoding())
+            .setMetadata(encodeMetadata(createOptions.getMetadata()))
+            .build();
+    return blobInfo;
+  }
+
+  private WriteChannel getClientLibraryWriteChannel(
+      Storage storage,
+      StorageResourceId resourceId,
+      CreateObjectOptions createOptions,
+      GoogleCloudStorageOptions storageOptions) {
+    WriteChannel writeChannel =
+        storage.writer(getBlobInfo(resourceId, createOptions), generateWriteOptions());
+    writeChannel.setChunkSize(
+        Math.toIntExact(
+            storageOptions.getWriteChannelOptions().getNumberOfBufferedRequests()
+                * MAX_WRITE_CHUNK_BYTES.getNumber()));
+
+    return writeChannel;
+  }
+
+  private class UploadOperation implements Callable<Boolean> {
+
+    // Read end of the pipe.
+    private final BufferedInputStream pipeSource;
+    private final StorageResourceId resourceId;
+    private final int MAX_BYTES_PER_MESSAGE = MAX_WRITE_CHUNK_BYTES.getNumber();
+
+    UploadOperation(InputStream pipeSource, StorageResourceId resourceId) {
+      this.resourceId = resourceId;
+      this.pipeSource = new BufferedInputStream(pipeSource, MAX_BYTES_PER_MESSAGE);
+    }
+
+    @Override
+    public Boolean call() throws Exception {
+      // Try-with-resource will close this end of the pipe so that
+      // the writer at the other end will not hang indefinitely.
+      try (pipeSource) {
+        boolean lastChunk = false;
+        ByteBuffer byteBuffer = ByteBuffer.allocate(MAX_BYTES_PER_MESSAGE);
+        while (!lastChunk) {
+          int remainingCapacity = byteBuffer.remaining();
+          ByteString data =
+              ByteString.readFrom(
+                  ByteStreams.limit(pipeSource, remainingCapacity), remainingCapacity);
+          if (data.size() < remainingCapacity) {
+            lastChunk = true;
+          }
+          byteBuffer.put(data.toByteArray());
+          // switch to read mode
+          byteBuffer.flip();
+          // this could result into partial write
+          writeToGcs(byteBuffer);
+          if (!lastChunk) {
+            // compact buffer for further writing
+            byteBuffer.compact();
+          }
+        }
+        // last chunk could be partially written
+        // uploading all bytes of last chunk
+        if (lastChunk && byteBuffer.hasRemaining()) {
+          while (byteBuffer.hasRemaining()) {
+            writeToGcs(byteBuffer);
+          }
+        }
+        return true;
+      } catch (Exception e) {
+        logger.atSevere().withCause(e).log("Exception while writing to channel");
+        throw new IOException(
+            String.format("Error occurred while uploading resource %s", resourceId), e);
+      }
+    }
+  }
+
+  private BlobWriteOption[] generateWriteOptions() {
+    List<BlobWriteOption> writeOptions = new ArrayList<>();
+
+    writeOptions.add(BlobWriteOption.disableGzipContent());
```

Comment:
This should be configurable via write/storage GCSIO options?

Reply:
Ideally it should be. As of now just replicated the behaviour from existing implementation https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java#L1277